compute_endpoint/globus_compute_endpoint/endpoint/rabbit_mq/result_publisher.py

```diff
@@ -0,0 +1,360 @@
+        assert self._mq_conn is not None, "Strictly called _by_ ioloop"
+        log.debug("%r Connection closed: %s", self, exc)
+        self.status = RabbitPublisherStatus.closed
+        self._mq_conn.ioloop.stop()
+
+    def _on_connection_open(self, _mq_conn: pika.BaseConnection):
+        log.debug("%r Connection established; creating channel", self)
+        self._open_channel()
+
+    def _open_channel(self):
+        if self._mq_conn.is_open:
+            self._mq_conn.channel(on_open_callback=self._on_channel_open)
+
+    def _on_channel_open(self, mq_chan: Channel):
+        self._mq_chan = mq_chan
+        self._delivery_tag_index = 0
+
+        mq_chan.add_on_close_callback(self._on_channel_closed)
+        mq_chan.add_on_cancel_callback(self._on_consumer_cancelled)
+
+        log.debug(
+            "Channel %s opened (%s); testing.",
+            mq_chan.channel_number,
+            mq_chan.connection.params,
+        )
+        mq_chan.exchange_declare(
+            passive=True,
+            exchange=self.queue_info["exchange"],
+            callback=self._on_exchange_verified,
+        )
+
+    def _on_consumer_cancelled(self, frame: Method[Basic.CancelOk]):
+        log.info("Consumer cancelled remotely, shutting down: %r", frame)
+        if self._mq_chan:
+            self._mq_chan.close()
+
+    def _on_channel_closed(self, mq_chan: Channel, exc: Exception):
+        self.status = RabbitPublisherStatus.closed
+        self._connected_at = None
+        now = time.monotonic()
+        then = now - self.channel_close_window_s
+        self._channel_closes = [cc for cc in self._channel_closes if cc > then]
+        self._channel_closes.append(now)
+        if len(self._channel_closes) < self.channel_close_window_limit:
+            if self._stop_event.is_set():
+                return
+            msg = f"{self} Channel closed  [{mq_chan}\n  ({exc})]"
+            log.debug(msg, exc_info=exc)
+            log.warning(msg)
+            mq_chan.connection.ioloop.call_later(1, self._open_channel)
+
+        else:
+            log.error(
+                f"{self} Unable to sustain channel after {len(self._channel_closes)}"
+                f" attempts in {self.channel_close_window_limit} seconds. ({exc})"
+            )
+            self._stop_event.set()
+
+    def _on_exchange_verified(self, _frame: Method[Exchange.DeclareOk]):
+        assert self._mq_chan is not None
+        self._mq_chan.queue_declare(
+            passive=True,
+            queue=self.queue_info["queue"],
+            callback=self._on_queue_verified,
+        )
+
+    def _on_queue_verified(self, _frame: Method[pikaQueue.DeclareOk]):
+        assert self._mq_chan is not None
+        self._mq_chan.confirm_delivery(self._on_delivery)
+        self.status = RabbitPublisherStatus.connected
+        self._connected_at = time.time()
+
+    def _on_delivery(self, frame: Method):
```

Comment:
Please add a docstring that `_on_delivery` will be triggered not just for single messages but for multiple.